### PR TITLE
Implement generic context

### PR DESCRIPTION
**Pull request by chingor13**
See #56

Using this generic context implementation is a much bigger PR, so lets' handle them separately.

### CODE REVIEW

File: src/Core/Context.php [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    public static function background()
+    {
+        return new Context();
+    }
+
+    /**
+     * Resets the context to an initial state. This is generally used only for
+     * testing.
+     */
+    public static function reset()
```
Comment (bogdandrutu) [inline src/Core/Context.php]: Is it any way we can make this not public but still available for testing?

File: src/Core/Context.php [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    public static function current()
+    {
+        if (!self::$current) {
+            self::setCurrent(self::background());
+        }
+
+        return self::$current;
+    }
+
+    public static function setCurrent(Context $toAttach)
```
Comment (bogdandrutu) [inline src/Core/Context.php]: I would not expose this for the moment. Users should use detach.

File: src/Core/Context.php [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    /**
+     * Reverses an attach(), restoring the previous context and exiting the
+     * current scope.
+     *
+     * @param  Context $toAttach
+     */
+    public function detach(Context $toAttach)
+    {
+        if (self::current() !== $this) {
+            trigger_error('Unexpected context to detach.', E_USER_WARNING);
```
Comment (bogdandrutu) [inline src/Core/Context.php]: Does this exit the function? We should just log the warning but still set the context if possible.

File: src/Core/Context.php [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+        self::setCurrent($toAttach);
+    }
+
+    /**
+     * Returns all the contained data.
+     *
+     * @return array
+     */
+    public function values()
```
Comment (bogdandrutu) [inline src/Core/Context.php]: Does this need to be public?

File: src/Core/Context.php [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    }
+
+    /**
+     * Creates a new context with the given key/value set.
+     *
+     * @param string $key
+     * @param mixed $value
+     * @return Context
+     */
+    public function withValue($key, $value)
```
Comment (tmatsuo) [inline src/Core/Context.php]: Do you need to call `withValue` multiple times if you want to add multiple key value pairs?

File: src/Core/Context.php (hunk 3 above)
Comment (tmatsuo) [reply src/Core/Context.php]: It depends on the runtime setting. On production, it only logs warnings.

File: src/Core/Context.php (hunk 2 above)
Comment (chingor13) [reply src/Core/Context.php]: Yeah, it's not really needed.

File: src/Core/Context.php (hunk 1 above)
Comment (chingor13) [reply src/Core/Context.php]: Unfortunately, not really. I can add an `@intenral` annotation, but that's only for documentation (i.e. some doc generators may skip it).

File: src/Core/Context.php (hunk 4 above)
Comment (chingor13) [reply src/Core/Context.php]: At some point, you may want all the values in the context. There's also the `value($key)` function to fetch a single entry.

File: src/Core/Context.php (hunk 5 above)
Comment (chingor13) [reply src/Core/Context.php]: Currently, yes. I can add a `withValues($keyValuesToMerge)`.

Comment (chingor13) [reply src/Core/Context.php]: Added.

**Pull request merged**